labs/b8.md

```diff
@@ -0,0 +1,84 @@
+
+#### GPG Keyring Abstraction
+GPG uses a "keyring" as a centralized location to hold all of a user's keys. You'll need to add/import a key to your keyring if you wish to use it and keep it around. Similarly, if you wish to share a key with someone else, you can export your key (which makes a copy of your key) and have them import it to their keyring. 
+
+To try it out:
+1. `gpg --full-generate-key` to generate a GPG public-private key pair. It'll ask for a password.
+2. `gpg -r [RECIPIENT] -e [FILE]` which'll encrypt `[FILE]` with `[RECIPIENT]`'s public key (for now, try encrypting a file with your own public key). 
+3. `gpg -d [FILE].gpg` will search through your keyring and decrypt the file with the appropriate private key (if you possess the correct private key, of course). 
+*You don't need to specify which key to decrypt a file with because GPG-encrypted files and keys contain metadata that allow GPG to pick the correct key from the keyring to decrypt the file with.*
+
+### Signatures
+The asymmetric scheme involving encryption with public key and decryption with private key can also be reversed to implement signatures, which allows for the authenticity of some data to be verified. In this reversed scheme, the private key is used to sign a file, producing a signature on that file. And the corresponding public key is used to verify the signature. 
+
+![alt](https://cdn-images-1.medium.com/max/1280/1*313akmm3vKOhK0bVNMxtSQ.png =400x250)
+
+To try it out:
+`gpg --sign [FILE]` to sign `[FILE]` with your private key. `gpg --verify [FILE].gpg` to verify that the file was signed by one of public keys on your keyring. 
+
+## Hashing (Checksums)
+Hash functions deterministically map arbitrary-length data to a fixed-length hash (in other words, a fixed number of bytes). As a result, the latter can serve as a summary of the former if the former exceeds the latter in terms of byte length. For instance, if we download a 1GB file and want to verify its integrity, instead of re-downloading the entire file again, we can simply compute a 20B hash of the file on our end and compare it to the 20B hash of source.
+
+To try it out:
+1. `sha1sum [FILE]` to get the SHA1 hash of `[FILE]`.
+2. `md5sum [FILE]` to get the MD5 hash of `[FILE]`.
+
+There are [many hash functions](https://en.wikipedia.org/wiki/List_of_hash_functions), some of which are cryptographic hash functions and others not. Crytographic hash functions primarily differ from non-cryptographic hash functions in that they are difficult to forge. For instance, if someone has a hash on hand, it is difficult for them to create a pre-hash file that maps to the same hash. If you are interested in all of the properties of a cryptographic hash function, [read here](https://en.wikipedia.org/wiki/Cryptographic_hash_function). In particular, SHA1 and MD5 have been proven to no longer be cryptographically secure and are only used for checksums to ensure data integrity.
```

Comment:
some of which are cryptographic hash functions and others not -> only some of which satisfy the requirements of cryptographic hash functions

or something like that

I'd also briefly explain what "difficult to forge" means, though that wording seems suspect in the context of hash functions